### DISCUSSION TITLE
ld: library not found for -lhwloc

### DISCUSSION THREAD

**Discussion by indexroot** [filecoin-project/lotus#5557] [General Q&A] [answered]
**Describe the bug**

**To Reproduce**
1. Run 
```
brew install go bzr jq pkg-config rustup hwloc
git clone https://github.com/filecoin-project/lotus.git
cd lotus/
export CGO_CFLAGS_ALLOW="-D__BLST_PORTABLE__"
export CGO_CFLAGS="-D__BLST_PORTABLE__"
make clean && make all
```

2. See error

**Expected behavior**
```
+ exit
rm -f lotus
go build  -ldflags="-X=github.com/filecoin-project/lotus/build.CurrentCommit=+git.6de62411c -extldflags=-L/usr/local/opt/openssl@1.1/lib" -o lotus ./cmd/lotus
# github.com/filecoin-project/filecoin-ffi/generated
ld: library not found for -lhwloc
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [lotus] Error 2
```

**Screenshots**
![image](https://user-images.githubusercontent.com/13267901/107201296-c1122400-6a33-11eb-92fc-8b8e658f8833.png)

**Version (run `lotus version`):**
latest

**Additional context**
~/.zshrc
Add  
```
export LDFLAGS="-L/usr/local/Cellar/hwloc/2.4.0/lib"
export CPPFLAGS="-I/usr/local/Cellar/hwloc/2.4.0/include"
```

![image](https://user-images.githubusercontent.com/13267901/107201895-85c42500-6a34-11eb-963c-156d9668729b.png)

**Comment by NewMai** [comment, marked as answer]
try to install libhwloc-dev with command:

`sudo apt install libhwloc-dev -y`

**Reply by jleni** [reply]
It seems that a mac is being used (`brew`) not ubuntu/linux

**Comment by lonnylot** [comment]
@indexroot Have you solved this yet? I'm experiencing the same issue. Here is where I am so far:

hwloc is found if you first:

`export LIBRARY_PATH=/opt/homebrew/lib`

However, then I receive this: `ld: warning: ignoring file /Users/lonnykapelushnik/Development/lotus/extern/filecoin-ffi/libfilcrypto.a, building for macOS-arm64 but attempting to link with file built for macOS-x86_64`

**Reply by jleni** [reply]
I have the same problem. 
There seems to be a problem when using new M1-based Macs.

**Reply by travisperson** [reply]
By default lotus build process downloads the Filecoin FFI build from Github. You will need to build it yourself here.

You can do so by setting the environment var `FFI_BUILD_FROM_SOURCE=1` prior to running the build. You may also need/want to set `RUSTFLAGS="-C target-cpu=native"`.

**Reply by lonnylot** [reply]
I wanted to add a message I sent in Slack last night so everyone knows where I left off:

> I then tried getting it to build libfilcrypto from source via:

```bash
export FFI_BUILD_FROM_SOURCE=1
brew install rust
ln -s /opt/homebrew/bin/rustup-init /opt/homebrew/bin/rustup
```

> Remove toolchain from `cargo` command in `rust/scripts/build-release.sh` because `nightly-2020-10-05` does not exist for `aarch64-apple-darwin`

> Which lead me to my current issue, which is sha2-asm doesn't work on the M1: https://github.com/RustCrypto/asm-hashes/issues/28 

I'm a bit busy today, but will try to add `RUSTFLAGS="-C target-cpu=native"` either later tonight or tomorrow night (GMT-5) and then, if it doesn't work, I'll continue down the rabbit hole.

**Reply by lonnylot** [reply]
Just a quick update: `RUSTFLAGS="-C target-cpu=native"` didn't change anything. Same issue as before.

**Comment by cryptonemo** [comment]
I wouldn't expect it to work out of the box.  At the very least, FFI does some CPU flag detection and I believe is limited to x86_64 and aarch at the moment.  I don't expect those changes will be very difficult, but something to look at if you're trying this route before it's officially tested/supported.

**Comment by jennijuju** [comment]
Issue can be tracked [here](https://github.com/filecoin-project/filecoin-ffi/issues/168)

**Reply by jennijuju** [reply]
related https://github.com/filecoin-project/lotus/issues/5979

**Reply by jennijuju** [reply]
https://github.com/filecoin-project/filecoin-ffi/pull/174

**Comment by jennijuju** [comment]
please use https://github.com/filecoin-project/filecoin-ffi/pull/174 and have `LIBRARY_PATH=/opt/homebrew/lib`

**Reply by jennijuju** [reply]
```
git clone https://github.com/filecoin-project/lotus
cd lotus
git submodule update --init --recursive
export GOARCH=arm64
export CGO_ENABLED=1
export LIBRARY_PATH=/opt/homebrew/lib
cd extern/filecoin-ffi
git fetch -a
git checkout m1-portable
make clean
make
cd ../../
make lotus
```

**Reply by sanderpick** [reply]
FYI, looks like `m1-portable` branch was [deleted](https://github.com/filecoin-project/filecoin-ffi/pull/174). I guess the changes are now in master?

**Reply by cryptonemo** [reply]
`rust-fil-proofs v7.0.1` contains the proper m1 support (not just a portable/compat mode).  I believe building lotus `v1.9.0-rc3` or later will have the support out of the box.

**Reply by sanderpick** [reply]
Cool, thanks!

**Reply by JustinGaoF** [reply]
@cryptonemo  Maybe I ask a stupid question. How can I use the rust-fil-proofs v7.0.1 while build the lotus?  I tried to use master and release/v1.9.0 branch of the lotus with master branch of filecoin-ffi, still met the issue `src/lotus/extern/filecoin-ffi/libfilcrypto.a, building for macOS-arm64 but attempting to link with file built for macOS-x86_64`

**Comment by pefish** [comment]
Is there any progress?

**Reply by cryptonemo** [reply]
Did you try: `export LIBRARY_PATH=/opt/homebrew/lib`?  The other option is to disable multicore sdr (see README: https://github.com/filecoin-project/filecoin-ffi)

**Comment by LexLuthr** [comment]
I was running into same error on M1 Mac with latest release when I was installing devnet. The following steps helped me.

1. brew install go bzr jq pkg-config rustup hwloc. - Please uninstall rust as it would clash with rustup in case you have installed
2. export LIBRARY_PATH=/opt/homebrew/lib
3. export FFI_BUILD_FROM_SOURCE=1
4. rustup-init (This should install cargo)
5. Follow the next steps as per the docs.
